### PR TITLE
Add resource fixtures services

### PR DESCRIPTION
**Pull request by loic425**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | 
| Related tickets | 
| License         | MIT

### CODE REVIEW

**Comment by loic425** [comment]
@lchrusciel Can you help me to add some phpUnit tests ?

File: src/Fixture/AbstractResourceFixture.php [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+use Doctrine\Common\Persistence\ObjectManager;
+use Sylius\Bundle\FixturesBundle\Fixture\Factory\ExampleFactoryInterface;
+use Sylius\Bundle\FixturesBundle\Fixture\FixtureInterface;
+use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
+use Symfony\Component\Config\Definition\Builder\TreeBuilder;
+use Symfony\Component\OptionsResolver\Options;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+abstract class AbstractResourceFixture implements FixtureInterface
```
Comment (lchrusciel) [inline src/Fixture/AbstractResourceFixture.php:24]: Do we even need to add `Resource` here? If I'm not missing something, this class could be used in any context, so just the creation of a sample structure for fixture and executing it in PHPunit would be enough(still not a no-brainer). What do you think, Loic?

Comment (loic425) [reply src/Fixture/AbstractResourceFixture.php:24]: You're right, we don't need Resource suffix. I will try a phpunit test tomorrow.. is there any similar tests on this bundle ?

Comment (lchrusciel) [reply src/Fixture/AbstractResourceFixture.php:24]: Not yet. These are closest I suppose: https://github.com/Sylius/SyliusFixturesBundle/tree/master/tests/Loader

Comment (loic425) [reply src/Fixture/AbstractResourceFixture.php:24]: @lchrusciel AbstractFixture already exists and it's a minimalist version without any example factory.

File: src/Fixture/AbstractResourceFixture.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Fixture;
```
Comment (lchrusciel) [inline src/Fixture/AbstractResourceFixture.php:14]: Namespace needs to be adjusted

File: src/Fixture/AbstractResourceFixture.php (hunk 1 above)
Comment (lchrusciel) [reply src/Fixture/AbstractResourceFixture.php:24]: What do you think about renaming it to `AbstractExampleFactoryAwareFixture`?

Comment (loic425) [reply src/Fixture/AbstractResourceFixture.php]: @lchrusciel I renamed it, I think this is a better name.
I don't know when I will add some phpUnit tests yet.

Comment (lchrusciel) [reply src/Fixture/AbstractResourceFixture.php]: Don't worry. Maybe someone will pick it up

File: src/Fixture/AbstractExampleFactoryAwareFixture.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    protected function getExampleFactoryClass(): string
+    {
+        $names = explode('\\', static::class);
+        $name = end($names);
+        $basename = preg_replace('/(.+)Fixture$/', '${1}', $name);
+
+        $pos = strrpos(static::class, '\\');
+        $namespace = false === $pos ? '' : substr(static::class, 0, $pos);
+
+        return $namespace.'\\Factory\\'.$basename.'ExampleFactory';
```
Comment (loic425) [inline src/Fixture/AbstractExampleFactoryAwareFixture.php:126]: @lchrusciel Maybe can we use $namespace.$basename.'ExampleFactory' instead to have a simpler convention?

File: src/Fixture/AbstractResourceFixture.php (hunk 1 above)
Comment (loic425) [reply src/Fixture/AbstractResourceFixture.php]: @lchrusciel Cause it's a little too long, I'm afraid developers doesn't remember which abstract service he should extend.
What do you think about `AbstractConfigurableFixture`?

**Comment by loic425** [comment]
@pamil @lchrusciel 
I've seen you worked recently a lot on this bundle, can we finish this one please?

File: src/Fixture/AbstractExampleFactoryAwareFixture.php [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    {
+        // empty
+    }
+
+    private function createExampleFactory(): ?object
+    {
+        $class = $this->getExampleFactoryClass();
+
+        return class_exists($class) ? new $class() : null;
+    }
```
Comment (lchrusciel) [inline src/Fixture/AbstractExampleFactoryAwareFixture.php:145]: How strongly are bounded to this solution? I'm not strongly against this, but my preferences are to keep away from putting too much logic (especially about dependencies wiring up) from class and inheritance as an extending option. This class will be a service anyway, so I would prefer to go with the autowiring/autoconfiguration way to make it easily accessible for new commers. For now, I would recommend the removal of this semi-magical dependency injection(creation?). 

However, if there will be a willingness in merging the current solution, the problem that I have (despite what I've written above) is that we have only one(pretty neat, I must say 🎉 ) test for this behaviour. My recommendations would be to test `getExampleFactoryClass` should be tested in separation, write additional cases for lack of factories at all, lack of interface in class and for the case, where we are injecting separate factory service.

**Review by lchrusciel — COMMENTED** [review]
Hey Loic,

thanks a lot for bootstrapping the testing environment for Fixtures and introducing default fixture and factory implementation.

File: tests/Application/config/config.yml [hunk 5]
```diff
@@ -1,13 +1,41 @@
...
+    suites:
+        default:
+            listeners:
+                orm_purger: null
+            fixtures:
+                dummy:
+                    options:
+                        random: 10
+                        custom:
+                            - foo: 'fighters'
```
[the remark this answers is not in the document]
Comment (loic425) [reply tests/Application/config/config.yml:32]: Yes I was inspired for this one :)

File: src/Fixture/AbstractExampleFactoryAwareFixture.php (hunk 4 above)
Comment (loic425) [reply src/Fixture/AbstractExampleFactoryAwareFixture.php:145]: Yes I think you're right. But I'm searching for a solution to have all autoconfigured without any service configuration to do for the developer.

Comment (loic425) [reply src/Fixture/AbstractExampleFactoryAwareFixture.php:145]: I will think about that next week

Comment (loic425) [reply src/Fixture/AbstractExampleFactoryAwareFixture.php:145]: @lchrusciel It was inspired by https://symfony.com/doc/current/validation/custom_constraint.html#creating-the-validator-itself.
I added some examples but maybe this is not a good idea like this...

Comment (loic425) [reply src/Fixture/AbstractExampleFactoryAwareFixture.php:145]: I've just think about this, you often have a resource with just simple properties and faker is enough to make your example factory. In other cases, you just need the entity manager and no other services (using getRepository($className) method. On Sylius world, that's not possible cause you need to use factories and repositories, but in a Symfony project, I think the entity manager is enough maybe all the time.

Comment (loic425) [reply src/Fixture/AbstractExampleFactoryAwareFixture.php:145]: But for Sylius case, the current way still work anyway (via the constructor). On Monofony projects, I override the constructor of the fixture service and specify typehint with the right example factory to use autowiring system.

**Pull request closed without merging**